docs/rfcs/functional-composition-pattern.md

```diff
@@ -0,0 +1,311 @@
+number_, because public interface types are always provided through a
+package-provided implementation.
+
+As a key requirement, every function must have a simple "no-op"
+implementation corresponding with the zero value of the
+`<Method>Func`. The expression `New<Type>(nil, nil, ...)` is the empty
+implementation for each type.
+
+## Key concepts
+
+### 1. Decompose Interfaces into Function Types
+
+Instead of implementing interfaces directly on structs, we create
+function types for each method:
+
+```go
+// Interface definition
+type RateReservation interface {
+    WaitTime() time.Duration
+    Cancel()
+}
+
+// Function types for each method
+type WaitTimeFunc func() time.Duration
+type CancelFunc func()
+
+// Function types implement their corresponding methods
+func (f WaitTimeFunc) WaitTime() time.Duration {
+    if f == nil {
+        return 0 // No-op behavior
+    }
+    return f()
+}
+
+func (f CancelFunc) Cancel() {
+    if f == nil {
+        return // No-op behavior
+    }
+    f()
+}
+```
+
+Users of the [`net/http` package have seen this
+pattern](https://pkg.go.dev/net/http#HandlerFunc). `http.HandlerFunc`
+can be seen as a prototype for the Functional Composition pattern, in
+this case for HTTP servers. Interestingly, the single-method
+`http.RoundTripper` interface, representing the same interaction for
+HTTP clients, does not have a `RoundTripperFunc` in the base library
+(consequently, this codebase defines [it for testing middleware
+extensions](`../../extension/extensionmiddleware/extensionmiddlewaretest/nop.go`)). In
+this codebase, the pattern is applied extensively.
+
+### 2. Compose Function Types into Interface Implementations
+
+Create concrete implementations embedding the function type
+corresponding with each interface method:
+
+```go
+// A struct embedding a <Method>Func for each method.
+type rateReservationImpl struct {
+    WaitTimeFunc
+    CancelFunc
+}
+```
+
+This pattern applies even for single-method interfaces, where the
+`<Method>Func` would implement the interface, were it not sealed. 
+
+```go
+// Single-method interface type
+type RateLimiter interface {
+    ReserveRate(context.Context, int) RateReservation
+}
+
+// Single function type
+ReserveRateFunc func(context.Context, int) RateReservation
+
+func (f ReserveRateFunc) ReserveRate(ctx context.Context, value int) RateReservation {
+    if f == nil {
+        return rateReservationImpl{} // Composite no-op behavior
+    }
+    f(ctx, value)
+}
+
+// The matching concrete type.
+type rateLimiterImpl struct {
+    ReserveRateFunc
+}
+```
+
+### 3. Use Constructors for Interface Values
+
+Provide constructor functions rather than exposing concrete types. By
+default, each interface should provide a `func
+New<Type>(<Method1>Func, <Method2>Func, ...) Type` for all methods,
+using the concrete implementation struct. For example:
+
+```go
+func NewRateReservation(wf WaitTimeFunc, cf CancelFunc) RateReservation {
```

Comment:
See my comment, we need optional pattern here, since if we extend `RateReservation` interface we can add more "params" (not directly but options) to the `NewRateReservation`